src/ch/ch_monitor.c

```diff
@@ -971,10 +971,59 @@ static void virCHMonitorStopEventLoop(virCHMonitorPtr mon)
     g_atomic_int_set(&mon->event_loop_stop, 1);
 }
 
+#define VM_PING_RETRIES 5
+
+static int
+virCHMonitorPostInitialize(virCHMonitorPtr mon, virDomainObjPtr vm,
+                           virCHDriverPtr driver)
+{
+    g_autoptr(virCHDriverConfig) cfg = virCHDriverGetConfig(driver);
+    virCHDomainObjPrivatePtr priv = vm->privateData;
+    int pings = 0;
+    int rv;
+
+    if ((rv = virPidFileReadPath(priv->pidfile, &vm->pid)) < 0) {
+        virReportSystemError(-rv,
+                             _("Domain %s didn't show up"),
+                             vm->def->name);
+        return -1;
+    }
+    VIR_DEBUG("cloud-hypervisor vm=%p name=%s running with pid=%lld",
+              vm, vm->def->name, (long long)vm->pid);
+
+    mon->pid = vm->pid;
+
+    VIR_DEBUG("Writing early domain status to disk");
+    if (virDomainObjSave(vm, driver->xmlopt, cfg->stateDir) < 0)
```

Comment:
Do we need this early save? I could not find any change happening in xml or domain state between this and the following `virDomainObjSave`

Reply:
There isn't going to be a difference in content. I think we probably could set a reason or maybe different state between the two points if we decide to keep the ping though.

Reply:
I was looking into these states a bit deeper. Looks like each state has its own enum of reasons. Combining a state with another state's reason would create issues with some helper functions that converts state to string. So maybe use `VIR_DOMAIN_NOSTATE` and `VIR_DOMAIN_NOSTATE_UNKNOWN` ?

We have not yet started the booting process even after the PING. Since the boot will start only after we call `vm.boot` API, probably setting state to NOSTATE would be the best option I feel. This is why I was thinking probably one `Save` is needed.

There is a small gotcha with setting to NOSTATE during early monitor initialization. When we define a domain, that state is set to SHUTOFF. During start, if we set it to NOSTATE, we should also take care to set it back to SHUTOFF, otherwise `virsh list --all` will show that state as `no state` if initialization fails.

Thoughts?

Reply:
I guess it is down to if it is more helpful to have the vm with `NOSTATE` or `SHUTOFF`. I feel like the difference is probably more relevant to developers than users since only we would know the VM ping had an issue if it was stuck in `NOSTATE`. Might be useful enough if there are bug reports with `no state` and the vm isn't seeming to come up but not something I suspect will occur.

Let's just remove the save before the ping and just not have the vm show up in the list until after that is successful for now.

Reply:
Understood. Thanks.

But I think VM will show as SHUTDOWN even if we don't do the save there. The `define` logic does a save and hence VM will show as SHUTOFF. This is same as qemu behavior and hence we are at parity.

Will remove the first save call now